### ISSUE TITLE
Not getting any video when using --gpu-api=d3d11 and --dither=error-diffusion combined. 

### ISSUE THREAD

**Issue by xXedgelordXx** [mpv-player/mpv#6814]
### mpv version and platform
Windows 10 LTSC 2019(1809 build).
RX 480 8GB with 1.9.5.2 Drivers

mpv 0.29.0-353-g65b1c2d065
ffmpeg version: git-2019-07-13-f86f213a

### Reproduction steps
Play any video.

### Expected behavior
Accurate dithering of colors.

### Actual behavior
I just get a blue screen inside the mpv's window. 

### Log file
https://pastebin.com/raw/USTm2VJY

### .conf file used
https://pastebin.com/raw/yhJZn115

### It DOES work with the vulkan api tho, vulkan log file:
https://pastebin.com/raw/EmnnZEn3

![dithererror](https://user-images.githubusercontent.com/42882555/61586166-d01c4300-ab2a-11e9-91b1-00d9e3d6f706.png)

I can hear the audio but not video, just a blue screen.

**Comment by bjin** [comment]
`D3DCompile` failed on SPIRV-Cross translated HLSL:
~~~
[   3.323][e][vo/gpu/d3d11] C:\Windows\system32\Shader@0x00000000088C5F30(27,9-41): error X3663: thread sync operation found in varying flow control, consider reformulating your algorithm so all threads will hit the sync simultaneously
[   3.323][e][vo/gpu/d3d11] C:\Windows\system32\Shader@0x00000000088C5F30(83,5-65): error X3663: this variable dependent on potentially varying data: stage_input
~~~

I don't know much about HLSL in general, but I think it's a D3D compiler bug. From the error log, the compiler concluded that the `if ((0 <= _71) && (_71 < 1779))` statement is a "varying", which is true. However it won't make the main loop "varying", because both branch of the statement will end current iteration and start next with a new barrier call. Those barrier calls should be able to sync with other threads. However, it's possible that the redundant `continue` statements generated by SPIRV-Cross confused the D3D compiler.
~~~
[   3.324][e][vo/gpu/d3d11] [ 70]             continue;
[   3.324][e][vo/gpu/d3d11] [ 71]         }
[   3.324][e][vo/gpu/d3d11] [ 72]         else
[   3.324][e][vo/gpu/d3d11] [ 73]         {
[   3.324][e][vo/gpu/d3d11] [ 74]             continue;
[   3.324][e][vo/gpu/d3d11] [ 75]         }
[   3.324][e][vo/gpu/d3d11] [ 76]         continue;
~~~

**Comment by bjin** [comment]
What we can do:
1. Find a workaround, which will probably works for specific version of D3D compiler only.
2. Report and wait for Microsoft to fix their D3D compiler, which will likely take years if not forever.
3. use vulkan and blame Microsoft for that.

I will choose 3.

**Comment by X7SHaDe** [comment]
Can confirm that the blue screen issue also happens on Intel HD 4000. Windows 10 build 1903 using --gpu-api=d3d11.

**Comment by xXedgelordXx** [comment]
I can confirm that it also happens with nvidia.
I swapped the rx 480 with a gtx 980 ti.
I uninstalled the AMD drivers with DDU in secure mode and internet disconnected. 
Running the 431.60 nvidia drivers, same windows ltsc 2019 and the same .conf as my original post. 
I just get a blue screen, the audio plays just fine but there's no video or OSD.

**.log file:**
https://pastebin.com/raw/T5p2ru8U

**Comment by bjin** [comment]
in case anyone missed that, an easy workaround is using `gpu-api=opengl` or `gpu-api=vulkan` on Windows.

If this issue turns out to be really annoying without an easy fix, it's always possible to disable error diffusion for `gpu-api=d3d11`.

**Comment by bjin** [comment]
I tested the generated HLSL code with some [online tool](http://shader-playground.timjones.io) and could confirm it's a bug in Microsoft FXC (the legacy D3D compiler).

1. The generated HLSL code compiles fine with [the new Microsoft DXC compiler](https://github.com/Microsoft/DirectXShaderCompiler) (entrypoint=main,profile=cs_6_4)
2. It fails with legacy D3D compiler, with the following error (entrypoint=main,profile=cs_5_1)
```
D:\home\site\wwwroot\unknown(27,9-41): error X3663: thread sync operation found in varying flow control, consider reformulating your algorithm so all threads will hit the sync simultaneously
``` 
3. After removing those three redundant `continue` statements, the generated HLSL code now compiles fine with the legacy D3D compiler.

So there are two approaches we could follow to fix this issue properly.
1. Fix SPIRV Cross to remove those redundant `continue` statements. 
2. Migrate to use the new Microsoft DXC compiler. Not sure whether packing with a new DLL is enough, or code change is required.

@rossy any idea?

**Comment by X7SHaDe** [comment]
For the second approach:

The new compiler seems to have a [bridge](https://github.com/microsoft/DirectXShaderCompiler/wiki/D3DCompiler-DXC-Bridge) to forward the calls to the old compiler into the new one by compiling the bridge then rename it to d3dcompiler_47.dll. But older GPUs (like Intel HD 4000) don't have support for Shader Model 6.0 though. And use of the bridge in these GPUs will break shader support.

So I think this is a possible solution?

Have both the bridge, the new, and the old compiler. The old compiler (d3dcompiler_47.dll) will be in a separate folder inside mpv. So the users will have a choice if they so happen to have an old gpu. I'm not a programmer so I don't know if the process of choosing between the new and the old may be automated. (If shader model 6.0 is supported, use new. If not, use old)

**Comment by bjin** [comment]
It seems that system `d3dcompiler_47.dll` is [forced](https://github.com/mpv-player/mpv/blob/master/video/out/d3d11/ra_d3d11.c#L2270-L2274) for Windows 8.1 and newer (instead of mpv packed DLL), so I guess at least some code change is required.

**Comment by rossy** [comment]
I get this bug too. Unfortunately, I don't think DXC/dxcompiler.dll is an option for mpv at the moment, since I'm pretty sure Direct3D 11 only supports shader model 4.0-5.0. I don't know what the bridge DLL is for, but it's probably for old Direct3D 12 apps that were designed for shader model 5.1. I don't think it would work for Direct3D 11 apps.

Anyway, using the bridge DLL with mpv produces this output for me:
```
[vo/gpu/d3d11] D3DCompiler version: 1.4.0.2343
[vo/gpu/d3d11] Failed to create blit() vertex shader: The parameter is incorrect. (0x80070057)
[vo/gpu/d3d11] after destroy:
[vo/gpu/d3d11] 104: ID3D11Device::CreateTexture2D: Returning E_INVALIDARG, meaning invalid parameters were passed.
[vo/gpu/d3d11] 166: ID3D11Device::CreateVertexShader: Vertex Shader is corrupt or in an unrecognized format.
[vo/gpu/d3d11] 166: ID3D11Device::CreateVertexShader: Shader is corrupt or in an unrecognized format.
```

> It seems that system ``d3dcompiler_47.dll`` is forced for Windows 8.1 and newer (instead of mpv packed DLL), so I guess at least some code change is required.

Yep. That was because d3dcompiler_47.dll became an inbox component in Windows 8.1, so the version packaged with Windows is likely to be newer than any that happened to be hanging around in mpv's folder.

Reporting the issue to SPIRV-Cross is probably the best thing to do. (Thanks @bjin for making the issue.) One of the goals of SPIRV-Cross is to emit clean code that looks like it could have been written by a human, so emitting redundant continue statements seems like a bug.

**Comment by bjin** [comment]
Should be fixed by SPIRV-Cross master now.

@rossy thanks for testing

**Comment by rossy** [comment]
This should be fixed by KhronosGroup/SPIRV-Cross#1131
